src/irmin-tezos-utils/files.ml

```diff
@@ -0,0 +1,198 @@
+    let len = max_bytes_needed_to_discover_length in
+    Io.read_exn prefix buffer ~off:poff ~len;
+    let kind, len = decode_entry_header len buffer in
+    Io.read_exn prefix buffer ~off:poff ~len;
+    let entry =
+      match kind with
+      | Inode_v1_unstable | Inode_v1_stable | Commit_v1 -> assert false
+      | Dangling_parent_commit | Commit_v2 -> `Commit
+      | Contents -> `Contents
+      | Inode_v2_root | Inode_v2_nonroot -> `Inode
+    in
+    (* Fmt.epr "traverse_prefix_chunk poff:%a len:%d\n%!" Int63.pp poff len; *)
+    let acc, continue = on_entry acc poff len entry in
+    let poff = Int63.add poff (Int63.of_int len) in
+    if continue && poff < end_poff then
+      traverse_prefix_chunk prefix ~end_poff buffer on_entry acc ~poff
+    else acc
+
+  let traverse_prefix fm buffer on_entry acc =
+    match (File_manager.prefix fm, File_manager.mapping fm) with
+    | _, None | None, _ -> ()
+    | Some prefix, Some mapping ->
+        let acc = ref acc in
+        File_manager.Mapping_file.iter mapping
+          (fun ~off:chunk_off ~poff:chunk_poff ~len:chunk_length ->
+            (* Fmt.epr "on_chunk\n%!"; *)
+            let on_entry acc off len x =
+              (* Shift from [poff] to [off] *)
+              let off = Int63.sub off chunk_poff in
+              let off = Int63.add chunk_off off in
+              on_entry acc off len x
+            in
+            let end_poff = Int63.add chunk_poff (Int63.of_int chunk_length) in
+            let poff = chunk_poff in
+            acc :=
+              traverse_prefix_chunk prefix ~end_poff buffer on_entry !acc ~poff)
+        |> Errs.raise_if_error;
+        !acc
+
+  let fold :
+      File_manager.t ->
+      on_entry:('a -> Int63.t -> int -> entry -> 'a * bool) ->
+      'a ->
+      'a =
+   fun fm ~on_entry acc ->
+    let suffix = File_manager.suffix fm in
+    let pl = File_manager.Control.payload (File_manager.control fm) in
+    (* [suffix_length] and [suffix_start_offset] do not account for the dead
+       header. *)
+    (* TODO: Rename [pl.entry_offset_suffix_end] to [suffix_length] *)
+    let suffix_length = pl.entry_offset_suffix_end in
+    let suffix_dead_header_size, suffix_start_offset =
+      match pl.status with
+      | From_v1_v2_post_upgrade _ -> (Int63.of_int 16, Int63.zero)
+      | From_v3_no_gc_yet | From_v3_used_non_minimal_indexing_strategy ->
+          (Int63.zero, Int63.zero)
+      | From_v3_gced x -> (Int63.zero, x.entry_offset_suffix_start)
+      | T1 | T2 | T3 | T4 | T5 | T6 | T7 | T8 | T9 | T10 | T11 | T12 | T13 | T14
+      | T15 ->
+          assert false
+    in
+    let buffer = Bytes.create (4096 * 4096) in
+    let acc = traverse_prefix fm buffer on_entry acc in
+    let acc =
+      let poff = suffix_dead_header_size in
+      let end_poff = Int63.add suffix_length suffix_dead_header_size in
+      let on_entry acc off len x =
+        (* Shift from [poff] to [off] *)
```

Comment:
I think this logic is also included in the [dispatcher](https://github.com/mirage/irmin/blob/main/src/irmin-pack/unix/dispatcher.ml#L56). maybe you could use `dispatcher.read_exn` for reads in suffix/prefix ?

Reply:
@Ngoguey42 I'll let this to you as you coded it and have more insight on it than me

Reply:
The dispatcher only contains logic to do random access of entries, not traversal. This code is pretty different from the dispatcher. 

Ideally we would create some traversal code directly in the dispatcher (or around it).